kong/tools/wrpc.lua

```diff
@@ -0,0 +1,685 @@
+    rpc = {},
+  }
+  local service_fname = assert(proto_searchpath(service_name))
+  local proto_f = assert(io.open(service_fname))
+  local scope_name = ""
+
+  for line in proto_f:lines() do
+    local annotation = line:match("//%s*%+wrpc:%s*(.-)%s*$")
+    if annotation then
+      local nextline = proto_f:read("*l")
+      local keyword, identifier = nextline:match("^%s*(%a+)%s+(%w+)")
+      if keyword and identifier then
+
+        if keyword == "service" then
+          scope_name = identifier;
+
+        elseif keyword == "rpc" then
+          identifier = scope_name .. "." .. identifier
+        end
+
+        local type_annotations = annotations[keyword]
+        if type_annotations then
+          local tag_key, tag_value = annotation:match("^%s*(%S-)=(%S+)%s*$")
+          if tag_key and tag_value then
+            tag_value = tag_value
+            local tags = type_annotations[identifier] or {}
+            type_annotations[identifier] = tags
+            tags[tag_key] = tag_value
+          end
+        end
+      end
+    end
+  end
+  proto_f:close()
+
+  grpc.each_method(service_fname, function(_, srvc, mthd)
+    assert(srvc.name)
+    assert(mthd.name)
+    local rpc_name = srvc.name .. "." .. mthd.name
+
+    local service_id = assert(annotations.service[srvc.name] and annotations.service[srvc.name]["service-id"])
+    local rpc_id = assert(annotations.rpc[rpc_name] and annotations.rpc[rpc_name]["rpc-id"])
+    local rpc = {
+      name = rpc_name,
+      service_id = tonumber(service_id),
+      rpc_id = tonumber(rpc_id),
+      input_type = mthd.input_type,
+      output_type = mthd.output_type,
+    }
+    self.methods[service_id .. ":" .. rpc_id] = rpc
+    self.methods[rpc_name] = rpc
+  end, true)
+end
+
+--- returns the method defintion given either:
+--- pair of IDs (service, rpc) or
+--- rpc name as "<service_name>.<rpc_name>"
+function wrpc_service:get_method(srvc_id, rpc_id)
+  local rpc_name
+  if type(srvc_id) == "string" and rpc_id == nil then
+    rpc_name = srvc_id
+  else
+    rpc_name = tostring(srvc_id) .. ":" .. tostring(rpc_id)
+  end
+
+  return self.methods[rpc_name]
+end
+
+--- sets a service handler for the givern rpc method
+--- @param rpc_name string Full name of the rpc method
+--- @param handler function Function called to handle the rpc method.
+--- @param response_handler function Fallback function called to handle responses.
+function wrpc_service:set_handler(rpc_name, handler, response_handler)
+  local rpc = self:get_method(rpc_name)
+  if not rpc then
+    return nil, string.format("unknown method %q", rpc_name)
+  end
+
+  rpc.handler = handler
+  rpc.response_handler = response_handler
+  return rpc
+end
+
+
+--- Part of wrpc_peer:call()
+--- If calling the same method with the same args several times,
+--- (to the same or different peers), this method returns the
+--- invariant part, so it can be cached to reduce encoding overhead
+function wrpc_service:encode_args(name, ...)
```

Comment:
It seems that all call to `call` and `call_wait` uses a table to contain arguments.

Reply:
not sure what do you mean.   the specific case i had in mind is the CP sending the same config to several DP nodes.  instead of repeating the call with the unencoded config, it calls `service:encode_args()` once, and then `peer:send_encoded_call()` for each DP node

Reply:
I mean like this:
```lua
local resp, err = peer:call_wait("ConfigService.ReportMetadata", { plugins = self.plugins_list })
assert(peer:call("ConfigService.PingCP", { hash = hash }))
```
This comment is related to the comment below.